### ISSUE TITLE
String length limiting broken for double-width characters

### ISSUE THREAD

**Issue by Evertras** [Evertras/bubble-table#40]
The following string word-wraps incorrectly:

```
直立した恐竜のような身体と、尻尾の先端に常に燃えている炎が特徴。
```

This is likely because we're measuring rune width and not character width.

The fix should likely be in `table/strlimit.go`

**Comment by RobertKwiatkowski** [comment]
Hi Evertras, 
I would like to take a shot on this issue. However I'm pretty new to go. Can you please tell me how to reproduce this issue?

**Comment by Evertras** [comment]
Hello, and welcome to Go! 🎊 

I think adding the test in strlimit_test.go is definitely ideal, but the issue isn't necessarily the `len` or rune count of the string.  The issue is that Chinese/Japanese characters (and others I'm sure) actually take up double the visual width as a Latin alphabet character.

To see what the issue is visibly, copy the string in the original issue above into one of the strings here, such as the name: https://github.com/Evertras/bubble-table/blob/main/examples/pokemon/main.go#L77

<img width="631" alt="image" src="https://user-images.githubusercontent.com/5923958/157145269-16699ba9-2cbb-4d44-bf44-519fe87befe0.png">

The `strlimit` function currently assumes that all characters have exactly the same width.  Unfortunately this is not the case, which causes the above issue.

We are already including https://github.com/mattn/go-runewidth as a dependency, since https://github.com/charmbracelet/lipgloss already requires it, so I think we should try to use this to calculate the total string width rather than the rune count.  Relevant reference docs here: https://pkg.go.dev/github.com/mattn/go-runewidth

Let me know if you need any more info, happy to help!

**Comment by Evertras** [comment]
Ah, and before I forget, this should be uncommented and work properly after this is fixed: https://github.com/Evertras/bubble-table/blob/main/examples/flex/main.go#L42

Also good for verifying the fix!

**Comment by RobertKwiatkowski** [comment]
Thank you for the nice description and all the sources. I tried to use the go-runewidth, however there is small problem about using odd numers. Example test:
```
		{
			name:     "Unicode chinese",
			input:    "直立した恐竜のような身体と、尻尾の先端に常に燃えている炎が特徴。",
			max:      3,
			expected: "直…",
		},
```
```

=== RUN   TestLimitStrFix/Unicode_chinese
    strlimitfix_test.go:83:
                Error Trace:    strlimitfix_test.go:83
                Error:          Not equal:
                                expected: "直…"
                                actual  : "\xe7\x9b…"

                                Diff:
                                --- Expected
                                +++ Actual
                                @@ -1 +1 @@
                                -直…
                                +��…
```

Do you have any idea how to solve truncation of the unicode?

**Comment by RobertKwiatkowski** [comment]
Thank you for this detailed instruction. I've created fork with the changes:
https://github.com/RobertKwiatkowski/bubble-table/commit/f98c7a42a7cdeb17c0b8a979ea7ba163ab857380
Please take a look at it, I'm not sure if it's clear solution because we have to keep in mind that every chinese char is equal to 2 instead of 1 in terms of length. The good point is that my solution passes all of your existing tests 😜

**Comment by Evertras** [comment]
Great, thanks!  I think this is on the right track and the tests look good.  For the record, though, that's Japanese, not Chinese. :)

Are you familiar with the process of opening a PR?  I think you can replace the existing `strlimit` code with what you have here (removing all the `fmt` lines and the 'before' comment) and create a PR.  Let me know if you need more info on how to do that, I'm just asking as I see your account is new. <3

> I've run this with both my solution and the existing code. However it looks exactly the same:

Hold down the left arrow key for a bit to shrink it.  If it suddenly starts word wrapping, it's broken.  If it properly uses the `...` character, then it's good!

**Comment by RobertKwiatkowski** [comment]
I believe it's working fine since it doesn't create second row of text but immediately uses the "..." after string is too long.
Here is my basic "manual" test:
https://user-images.githubusercontent.com/101144854/157246112-e760f37a-5f9d-452a-a11b-194c917e9185.mp4

I'm going to create the pull request.
Don't worry, I'm relatively new to go, but I have some C++ experience 😜

**Comment by Evertras** [comment]
Manual test LGTM!

**Comment by RobertKwiatkowski** [comment]
Created pull request, please let me know if I should fix anything related to the commit format or code.

**Comment by Evertras** [comment]
LGTM and released as v0.8.2, thanks again!
